Add LogMonkey Barrel

KenConnectIQ opened the pull request:
Adds a LogMonkey barrel project that contains logging utilities
useful for debugging.

Commit:
Add LogMonkey Barrel

Adds a LogMonkey barrel project that contains logging utilities
useful for debugging.

Change-Id: I62b6962c8fda05bddd2d4e39f1bdd9df5145a773

Commit:
Add LogMonkey Barrel

Adds a LogMonkey barrel project that contains logging utilities
useful for debugging.

Change-Id: I62b6962c8fda05bddd2d4e39f1bdd9df5145a773

blaskovicz commented:
If this existed I probably could have avoided writing my own logger!

https://github.com/blaskovicz/garmin-nest-camera-control/blob/master/source/Logger.mc

blaskovicz commented:
@KenConnectIQ the link in your readme to 'monkey barrels' on the connect-iq developer guide doesn't seems dead; are you intending a new guide to be added there as well?
https://developer.garmin.com/connect-iq/programmers-guide/monkey-barrels

KenConnectIQ commented:
I tried to update the URLs in the links in this change to where they will be before pushing. The Monkey Barrels feature is in beta right now and the documentation on the developer website reflects the current production release documentation.

I'm glad to see there's some interest in a logging tool though. I'm hoping we see enough interest in this tool to evolve and improve it over time.

KenConnectIQ commented on barrels/LogMonkey/LogTests.mc:
```diff
@@ -0,0 +1,246 @@
+        //! This class mocks the Toybox.System.println() interface but instead
+        //! of printing a value saves it to a string value which can be retrieved
+        //! and examed to verify log results are correct.
+        class MockSys {
```
Unless we use the initialize function to set mStreamContents to an empty string as suggested below - which is ultimately what I did - I don't think this is necessary. This class isn't overriding another class so there's not a parent initialize to call and the compiler will insert an empty initialize function if you don't include one so any class that would subclass this one would be able to call the initialize symbol safely.

KenConnectIQ commented on barrels/LogMonkey/LogTests.mc:
```diff
@@ -0,0 +1,246 @@
+            //! A mock of Toybox.System.println() function
+            //! @param [Toybox::Lang::Object] message The message to print/log
+            public function println(message) {
+                if (mStreamContents == null) {
```
Done

KenConnectIQ commented on barrels/LogMonkey/LogTests.mc:
```diff
@@ -0,0 +1,246 @@
+        class TestException extends Lang.Exception {
+            function initialize() {
+                Exception.initialize();
+                mMessage = "This is a test Exception";
```
Done

KenConnectIQ commented on barrels/LogMonkey/LogTests.mc:
```diff
@@ -0,0 +1,246 @@
+            log.logVariable(tag, varName, varToTest);
+            logOutput = mockSys.getStreamContents();
+            logger.debug("log output: " + logOutput);
+            Test.assert(logOutput.find("(Array)") != null);
```
I'd be open to looking for a more exact value if you have suggestions on how to do it. I really just wanted a test to ensure that when given a variable type to getVariableType() we'd get the expected value in parentheses in the output message. The problem I was running into is that the log format includes a timestamp so we can do an exact match and there's not a great way to do a regex type of check. We could have a function that breaks the message down and looks for a specific part to match exactly but I don't know if that will provide better value that just looking for the specific type piece.

KenConnectIQ replied:
I split up the test types unit test into the individual variable types using the helper function you suggested. I'm going to leave the mocked time values out as I don't want to have the unit test dictate the constructor interface for the Logger class.

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+    //! is set within the Logger.initialize() function.
+    class Logger {
+
+        private var mLogLevel, mLogStream;
```
Done

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+    //! is set within the Logger.initialize() function.
+    class Logger {
+
+        private var mLogLevel, mLogStream;
```
I like the idea of separating this by class. I'm going to take that approach.

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+        //! @param [Toybox::Lang::String] The log level value to log messages through this class to
+        //! @param [Toybox::Lang::Object] An object which defines a println(message) function
+        (:debug)
+        function initialize(logLevel, logStream) {
```
An earlier conversation with Nick resulted in restructuring the log classes with the idea that different implementations of the class could be swapped out/configured. I wanted to explicitly document what's expected as inputs to a Logger class so in the future if there's a configuration option that external developers could use to setup their own loggers or if someone internally is going in and modifying this barrel it's documented.

KenConnectIQ replied:
I don't think that's the case. The last parameter is simply stating you must provide a stream object that defines a println method. If better options for writing to a file/dedicated log get added in the future then that object can be swapped out as needed. I also think the idea was that in the future there could be some sort of setup/initialize function with LogMonkey that allows the user to define the underlying output/destination of the log which would then create the local variable objects.
```
module LogMonkey {
    var Debug;
    var Warn;
    var Error;

    function setup(logStream) {
        Debug = new Logger("D", logStream);
        Warn = new Logger("W", logStream);
        Error = new Logger("E", logStream);
    }
}
```
We're just not going to include any sort of setup now because the underlying system just doesn't support a more advanced infrastructure.

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+        // The string formats to use when printing log messages
+        private static var FORMAT_VARIABLE = "$1$=($2$) $3$";
+        private static var FORMAT_LOG_MESSAGE = "(lmf1)[$1$] {$2$} $3$: $4$";
+        private static var FORMAT_TIMESTAMP = "$1$-$2$-$3$ $4$:$5$:$6$";
```
These should be taken care of with the empty class approach discussed above.

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+            // Get a timestamp from the system
+            var currentTime = Greg.info(Time.now(), Time.FORMAT_SHORT);
+            var timestamp = Lang.format(FORMAT_TIMESTAMP, [
+                currentTime.month.format("%02u"),
```
I like this reasoning. I'll update the format.

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        (:release)
+        private function formLogMessage(tag, message) {
```
Done

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        (:release)
+        private function log(tag, message) {
```
Done

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        (:release)
+        private function getVariableType(variable) {
```
Done

KenConnectIQ commented on barrels/LogMonkey/Logger.mc:
```diff
@@ -0,0 +1,172 @@
+        //! @param tag The tag to apply to the log message
+        //! @param message The message to log
+        (:debug)
+        public function logMessage(tag, message) {
```
I think with the introduction of public/private/protected we should start using them as a coding convention even if it's not grammatically required but I'm also not the authority to dictate this. Unless there's a strong opinion against it I'm going to leave them in.

KenConnectIQ replied:
I agree it is a bit more Java-y but so is switching the grammar from hidden/<nothing> to private/protected/public. If we chose to go that route with the grammar I think it may make sense to update our teams convention approach. I'll update the patch to remove the public definitions and leave this open to debate internally.

KenConnectIQ commented on barrels/LogMonkey/README.md:
````diff
@@ -0,0 +1,77 @@
+```
+using LogMonkey as Log;
+```
+Within you're project's code you can log messages, variables and Exceptions via the available APIs.
````
Done

KenConnectIQ commented on barrels/LogMonkey/README.md:
```diff
@@ -0,0 +1,77 @@
+
+## Viewing and Parsing Logs
+
+LogMonkey makes use of `Toybox.System.println()` calls to log information. This means that if your app also makes use of `pinrtln()` calls they will show up alongside the log entries.
```
Done

KenConnectIQ commented on barrels/LogMonkey/README.md:
```diff
@@ -0,0 +1,77 @@
+
+### Output Format
+
+LogMonkey logs each line in the file in the following format: `(log format version)[timestamp] {log level} tag: message`. Details for each block of this format are given in the table below.
```
Done

KenConnectIQ commented on barrels/LogMonkey/README.md:
```diff
@@ -0,0 +1,77 @@
+| ------------------ | ------------------------------------------------------ | ----------------------------------------------------------------------------------- |
+| log format version | A value beginning with "lfm" then a number             | The output format version of the line in the log.                                   |
+| timestamp          | MM-DD-YYYY HH:MM:SS                                    | A time stamp of when the log entry occurred.                                        |
+| log level          | Regular Expression: `[a-zA-Z0-9_]`                     | The priority level of the log entry. LogMonkey supports values of `D`, `W` and `E`. |
```
The log level value should can be more than a single character; the regular expression here should be updated to reflect that. I wanted the standard format to be meaningful but concise so it's less overwhelming to view a raw log file.

KenConnectIQ commented on barrels/LogMonkey/README.md:
```diff
@@ -0,0 +1,77 @@
+| log format version | A value beginning with "lfm" then a number             | The output format version of the line in the log.                                   |
+| timestamp          | MM-DD-YYYY HH:MM:SS                                    | A time stamp of when the log entry occurred.                                        |
+| log level          | Regular Expression: `[a-zA-Z0-9_]`                     | The priority level of the log entry. LogMonkey supports values of `D`, `W` and `E`. |
+| tag                | A string containing any character except a colon (`:`) | A brief tag value that can be used to quickly filter and find specific log entries. |
```
I think I choose to use a regular expression above because it was easier - from my view - to understand it but I'm going to use plain English for both.

KenConnectIQ commented on barrels/LogMonkey/README.md:
```diff
@@ -0,0 +1,77 @@
+
+### Viewing Logs on a Device
+
+When running your app on a physical device the log statements get added to the app's log file. The app's log file can be viewed by plugging your device into your computer and navigating to `<device root>/GARMIN/APPS/LOGS`. The app log file will match the name of the app prg file (located in `<device root>/GARMIN/APPS`.
```
Probably. I'll add that.

KenConnectIQ replied:
I think I missed this when rebasing my change. The last commit I pushed adds this bit back in.

KenConnectIQ commented on barrels/README.md:
```diff
@@ -1,8 +1,25 @@
+### LogMonkey
+The [LogMonkey barrel](https://github.com/garmin/connectiq-apps/tree/master/barrels/LogMonkey)
+provides some basic logging utilities to aid in developing Connect IQ apps. Inclusions of the
+LogMonkey within an app will automatically be compiled out or minimized when building a release
```
I'm going to change the entire sentence to be: Including LogMonkey within an app and building the app in release mode will automatically minimize the compiled output to save run time memory.

KenConnectIQ commented:
I just have to push a new change to this branch and the pull request will get updated. I haven't yet pushed changes but should be doing so before lunch today.

Commit:
Address Rewiew Comments 1

Address Travis's review comments.

Change-Id: Ic075e04c9a01e930fc45e7e3ff962b6c9ec5346d

Commit:
Address Rebasing Error

Address a rebasing error I introduced while responding to Travis's
review comments.

Change-Id: I4e081708633a4a1a7e14d2aa1debd2e8e0f7aaec

Commit:
Address Review Comments 2

Address additional review comments left by Travis.

Change-Id: Ib67cd7ef720a0fc0fc34e3912f15d80c9d634c3a

KenConnectIQ commented on barrels/LogMonkey/README.md:
```diff
@@ -0,0 +1,78 @@
+| Block              | Format                                                    | Description                                                                         |
+| ------------------ | --------------------------------------------------------- | ----------------------------------------------------------------------------------- |
+| log format version | A value beginning with "lfm" then a number                | The output format version of the line in the log.                                   |
+| timestamp          | MM-DD-YYYY HH:MM:SS                                       | A time stamp of when the log entry occurred.                                        |
```
Missed this one. Done.
